zebra-network/src/protocol/external/codec.rs

```diff
@@ -124,8 +123,8 @@ impl Encoder for Codec {
             return Err(Parse("body length exceeded maximum size"));
         }
 
-        if let Some(addr) = self.builder.addr.as_ref() {
-            metrics::counter!("bytes.written", (body.len() + HEADER_LEN) as u64, "addr" =>  addr.ip().to_string());
+        if let Some(addr) = self.builder.addr.clone() {
```

Comment:
I tried to avoid this clones but i was not able to make it build other way. Any suggestion is appreciated.

Reply:
Yep, I think that cloning at some point will be required.  It looks like the macro will construct a [`Label`](https://docs.rs/metrics/0.12.1/metrics/struct.Label.html) from something that's `Into<Cow<'static, str>>`.  The copy-on-write [`Cow<'a, T>`](https://doc.rust-lang.org/nightly/alloc/borrow/enum.Cow.html) type can be either `Owned` or `Borrowed<'a>`.  The `'a` marker tracks how long the borrowed data is supposed to live. 

We know that we want something that's `Into<Cow<'static, str>>`.  `Into` is implemented automatically as the reverse of `From`.  So we are looking for `impl From</* something */> for Cow<'static, str>`.  Looking at the [implementations](https://doc.rust-lang.org/nightly/alloc/borrow/enum.Cow.html#trait-implementations) section, there are two relevant impls:

* [`impl<'a> From<&'a String> for Cow<'a, str>`](https://doc.rust-lang.org/nightly/alloc/borrow/enum.Cow.html#impl-From%3C%26%27a%20String%3E)

* [`impl<'a> From<String> for Cow<'a, str>`](https://doc.rust-lang.org/nightly/alloc/borrow/enum.Cow.html#impl-From%3CString%3E)

So, we can construct a `Cow<_, str>` from a borrowed string (first impl) or an owned string (second impl).  But if we construct it from a borrowed string with lifetime `'a` (aka `&'a String`), we end up with a `Cow<'a, str>`, marking that the `Cow` can't outlive the data it borrows.  On the other hand, when we construct it from an owned string, we get a `Cow<'a, str>` with no constraint on the lifetime `'a`, so `'a` can be *any* lifetime.  This is because the `Cow` owns the data, so it can live as long as it likes.

Looking at the `metrics` crate again, we need to have a `Cow<'static, str>`.  So we can't construct this from a borrowed `String` unless the borrowed `String` has a `'static` lifetime (can live forever).  But the string we have will only live as long as the codec, which does not live forever.  

This is all a very long way to say that we do have to do a clone here, but hopefully this shows an example of the process of working out from the type signatures that the clone is actually required.

Reply:
Thanks for the details, i was able to follow them and now i have at least some idea on what is going on here.

Reply:
We will do the clones here so marking the question as resolved. Thanks again @hdevalence